src/lib/snark_worker/functor.ml

```diff
@@ -104,15 +103,19 @@ module Make (Inputs : Intf.Inputs_intf) :
       with
       | Error e ->
           log_and_retry "getting work" e
-      | Ok None ->
+      | Ok (None, _) ->
           let random_delay =
             Worker_state.worker_wait_time
             +. (0.5 *. Random.float Worker_state.worker_wait_time)
           in
           (* No work to be done -- quietly take a brief nap *)
           let%bind () = wait ~sec:random_delay () in
           go ()
-      | Ok (Some work) -> (
+      | Ok (Some _work, None) ->
+          log_and_retry
```

Comment:
the problem with this branch is that when the daemon gives out work, it puts the work in a "seen" bucket and then it will not give that same piece of work out for 2 minutes! So as written, we'll be slowing down the work completion process too much by letting the daemon give us work without also giving us a key.